hashintel/hash
# [ENGINE] Adding Structured Logging

## 🌟 What is the purpose of this PR?

Fixes the JSON logging output, and adds structured logging to a file to the engine

## 🔍 What does this change?

- Fixes `json` as an `emit` option on the CLI and hash_engine binary args
- Adds a non-blocking file appender to the tracing subscriber, to write JSON-formatted logs to the file

### Does this require a change to the docs?


- [x] We should say where the log folder is and what's in it (TODO)

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1199548034582004/1201546500237308) (_internal_)

## 🛡 What tests cover this?



- None

## ❓ How to test this?



1.  Checkout the branch / view the deployment
2.  Run any experiment
3. Check the ./log directory
4. Also try with `--emit json` and confirm the stderr is in JSON formatting

## 📹 Demo

![image](https://user-images.githubusercontent.com/25749103/151041216-12b4d2e0-f88d-49cc-a5bf-5afa9fe9e42c.png)
![image](https://user-images.githubusercontent.com/25749103/151041286-0ef29b37-89b9-42f0-9613-406a22d7b02d.png)
